airbyte-cdk/python/airbyte_cdk/entrypoint.py

```diff
@@ -160,8 +161,36 @@ def read(
         if self.source.check_config_against_spec:
             self.validate_connection(source_spec, config)
 
-        yield from self.source.read(self.logger, config, catalog, state)
-        yield from self._emit_queued_messages(self.source)
+        stream_message_counter: DefaultDict[HashableStreamDescriptor, int] = defaultdict(int)
+        for message in self.source.read(self.logger, config, catalog, state):
+            yield self.handle_record_counts(message, stream_message_counter)
+        for message in self._emit_queued_messages(self.source):
+            yield self.handle_record_counts(message, stream_message_counter)
+
+    @staticmethod
+    def handle_record_counts(message: AirbyteMessage, stream_message_count: DefaultDict[HashableStreamDescriptor, int]) -> AirbyteMessage:
+        if message.type == Type.RECORD:
+            stream_message_count[message_utils.get_stream_descriptor(message)] += 1
+
+        elif message.type == Type.STATE:
+            if not message.state.stream:
+                raise ValueError("State message was not in per-stream state format, which is required for record counts.")
+
+            stream_descriptor = message_utils.get_stream_descriptor(message)
+
+            # Set record count from the counter onto the state message
+            message.state.sourceStats = message.state.sourceStats or AirbyteStateStats()
+            message.state.sourceStats.recordCount = stream_message_count.get(stream_descriptor, 0)
+
+            # TODO: Logging - remove after pre-release testing
```

Comment:
```suggestion
            # TODO: Logging for record counts. Once pre-release testing of record counts in state messages completes, this is safe to remove.
```

Reply:
I don't think we need to change this because we're removing this block anyway, pre-release testing was done this week. There is some risk to logging state messages because it's theoretically possible for PII and even a potentially very large state message (especially in concurrent could break it).